libc++ port to TizenRT

manoharahk opened the pull request:
Below series of patch set adds llvm libc++ support in TizenRT.

seinfra commented:
Code Rule Check OK.

sunghan-chang commented:
First commit is ```libxx : port missing cxx changes from latest nuttx```. But many changes are not cxx. Could you change a name approprietary?

sunghan-chang commented on apps/examples/helloxx/Makefile:
```diff
@@ -1,6 +1,6 @@
 ###########################################################################
 #
-# Copyright 2016 Samsung Electronics All Rights Reserved.
+# Copyright 2017 Samsung Electronics All Rights Reserved.
```
Don't change this.

manoharahk replied:
OK

sunghan-chang commented on apps/examples/helloxx/helloxx_main.cxx:
```diff
@@ -102,35 +105,35 @@
-  private:
-    int mSecret;
+public:
+	CHelloWorld(void) : mSecret(42)
```
You should split changing indentation from this commit.

sunghan-chang commented on apps/platform/gnu/gnu_cxxinitialize.c:
```diff
@@ -0,0 +1,147 @@
+ */
+
+#ifdef CONFIG_DEBUG_CXX
+#define cxxinfo        _info
```
Could you let me know where ```_info``` definition? Here are toooo many changes so that it is difficult to find where it is. Could you help me?

manoharahk replied:
_info is missing, It will changed to "vdbg"

sunghan-chang commented on build/configs/artik053/helloxx/Make.defs:
```diff
@@ -0,0 +1,161 @@
+LIBEXT = .a
+EXEEXT =
+
+ifneq ($(CROSSDEV),arm-nuttx-elf-)
```
Please remove this checking.

manoharahk replied:
Ok

sunghan-chang commented on build/configs/artik053/helloxx/Make.defs:
```diff
@@ -0,0 +1,161 @@
+  DIRLINK = $(TOPDIR)/tools/copydir.sh
+  DIRUNLINK = $(TOPDIR)/tools/unlink.sh
+  MKDEP = $(TOPDIR)/tools/mkwindeps.sh
+  ARCHINCLUDES = -I. -isystem "${shell cygpath -w $(TOPDIR)/include}"
```
Please add external on ```ARCHINCLUDES``` and ```ARCHXXINCLUDES``` for win and linux like others.

manoharahk replied:
OK

manoharahk commented:
@sunghan-chang ,
There are no TC's implemented for new libc changes ported. 
So should we change all of  these new API  to have below comment?
+/**
+ * @cond
+ * @internal
+ *

sunghan-chang commented:
@manoharahk You need to discuss about tc with Samir.
If you can finish making tc before first release, you can keep it with since v1.x.
But you can't, you should change it to internal as you said. (cond internal endcond)

sunghan-chang commented:
Could you merge cxxtest and helloxx commits? You can push them into a commit with mentioning of location like below:
```
artik053/cxxtest : add --gc-sections to linker for garbage collection 
artik053/helloxx : add --gc-sections to linker for garbage collection
==>
artik053/cxxtest, helloxx : add --gc-sections to linker for garbage collection
```

manoharahk commented:
Okay, add "add --gc-sections to linker" patches will be merged as requested in the next patch-set.

sunghan-chang commented:
I want to confirm that is an option. You can decide what is helpful to developer or meanless whatever.
If you will merge them, full optimization, removing linking supc and so on should be merged as well.

samirsahu1947 commented:
@sunghan-chang All APIs were tested with sample test programs which I referred in meeting last week. But they are not formal and cant be pushed. But definately we are planning to create TCs from the test programs before the first release. Pushing as internal doesnt make any sense keeping in mind the purpose of CPP layer

manoharahk commented:
Hello Mr.Chang & Samir,

The above patch-set address all the comments given above except making new API's internal.
Change Log:
   - Changed license to "Copyright 2018" in all newly added files
   - Updated helloxx & cxxtest Make.defs to include external directory
   - Squashed cxxtest and helloxx commits to relevant single commit
   - Split helloxx Makefile & indentation changes
  -  Updated cxxinfo macro

seinfra commented:
Code Rule Check OK.

pillip8282 commented on build/configs/artik053/cxxtest/Make.defs:
```diff
@@ -0,0 +1,155 @@
+  # Linux/Cygwin-native toolchain
+  MKDEP = $(TOPDIR)/tools/mkdeps$(HOSTEXEEXT)
+  ARCHINCLUDES = -I. -isystem $(TOPDIR)/include -isystem $(TOPDIR)/../framework/include -isystem $(TOPDIR)/../external/include
+  ARCHXXINCLUDES = -I. -isystem $(TOPDIR)/include -isystem $(TOPDIR)/include/cxx -isystem $(TOPDIR)/include/libcxx -isystem $(TOPDIR)/../framework/include -isystem $(TOPDIR)/../external/include
```
Does it need to include both 'include/cxx' and 'include/libcxx'?
Some files are located both folders for example cstdlib and cstring are.
And it was not built when I removed 'include/cxx'

sunghan-chang replied:
Could you let us know the reason that you make a new folder, libcxx in include?

manoharahk replied:
1. Libxx is tizenRT/nuttx's minimal core c++ implementation. include/cxx has the headers of this.
2. Libc++(libcxx) is llvm libc++ standard library. include/libcxx has headers of this.

It was done to separate libc++ from libxx. We need both & cant mix.

pillip8282 replied:
I think we should delete one of duplicated files which have same name in both folders. It is confusing

manoharahk replied:
I think applications need not worry, they just need to include.
Make defs already set the include path in the correct order.
It will search first include/cxx. If it wont find, then it will include from include/libcxx/

pillip8282 replied:
In that case, Application will refer cmath in include/cxx not cmath in include/libcxx even though User selected llvm libc++ library. Do you think that is it right?

manoharahk replied:
yes

pillip8282 commented on build/configs/artik053/cxxtest/Make.defs:
```diff
@@ -0,0 +1,155 @@
+CFLAGS = $(ARCHCFLAGS) $(ARCHWARNINGS) $(ARCHOPTIMIZATION) $(ARCHCPUFLAGS) $(ARCHINCLUDES) $(ARCHDEFINES) $(EXTRADEFINES) -pipe -ffunction-sections -fdata-sections
+CPICFLAGS = $(ARCHPICFLAGS) $(CFLAGS)
+CXXFLAGS  = $(ARCHCXXFLAGS) $(ARCHWARNINGSXX) $(ARCHOPTIMIZATION) $(ARCHCPUFLAGS) $(ARCHXXINCLUDES)
+CXXFLAGS += $(ARCHDEFINES) $(EXTRADEFINES) -pipe -std=c++11 -DCLOCK_MONOTONIC -D__TINYARA__
```
Defining -DCLOCK_MONOTONIC and -D_TINYARA_ at here is not good. for example You can set monotonic option by Kconfig.

manoharahk replied:
Not only these two, there are other "defines" as well  we need to pass to libcxx.
So in my opinion, adding all these defines/configs of libc++ into TizenRT Kconfig and including tinyara/config.h in all the libc++ source is unnecessary.

pillip8282 replied:
You can set -DCLOCK_MONOTONIC by Kconfig. and -D__TINYARA__ is defined Makefile of each modules. so I think that it would be better to set __TINYARA__ at Makefile in lib/libxx/.

pillip8282 commented on build/configs/artik053/cxxtest/Make.defs:
```diff
@@ -0,0 +1,155 @@
+CPICFLAGS = $(ARCHPICFLAGS) $(CFLAGS)
+CXXFLAGS  = $(ARCHCXXFLAGS) $(ARCHWARNINGSXX) $(ARCHOPTIMIZATION) $(ARCHCPUFLAGS) $(ARCHXXINCLUDES)
+CXXFLAGS += $(ARCHDEFINES) $(EXTRADEFINES) -pipe -std=c++11 -DCLOCK_MONOTONIC -D__TINYARA__
+CXXFLAGS += -fno-builtin -fno-exceptions -fcheck-new -fno-rtti
```
some options(ex. -fno-builtin) are set multiple times.
It would be better to set it once.

manoharahk replied:
I will check this

manoharahk replied:
Done

seinfra commented:
Code Rule Check OK.

manoharahk commented:
Removed "Public"  Doxygen comments for new libc API's added and marked them as internal

seinfra commented:
Code Rule Check OK.

manoharahk commented:
Change Log:
1. Removed include/cxx headers and updated include/libcxx
2. Updated cxxtest & helloxx Make.defs to remove cxx include path

manoharahk commented on lib/libxx/Kconfig:
```diff
@@ -38,6 +38,36 @@ config CXX_NEWLONG
+comment "LLVM C++ Library (libcxx)"
+
+config LIBCXX
+	bool "Build LLVM libcxx (must be installed)"
```
What i can understand is , install the source first. Since we are including libcxx source inside tizenrt we can can remove this.

sunghan-chang replied:
Same as you. LLVM will be inside TizenRT so that you need to remove that sentence.

manoharahk commented on lib/libxx/Kconfig:
```diff
@@ -38,6 +38,36 @@ config CXX_NEWLONG
+	bool "Build LLVM libcxx (must be installed)"
+	default n
+	---help---
+		If you have installed libcxx into the NuttX source try, then it can
```
Missed, i will remove this

seinfra commented:
Code Rule Check OK.

manoharahk commented on os/include/elf32.h:
```diff
@@ -0,0 +1,361 @@
+ *
+ ****************************************************************************/
+
+#ifndef __INCLUDE_ELF32_H
```
It was added to include in libxx__gnu_unwind_find_exidx.hxx for supporting __gnu_Unwind_Find_exidx .

It got included while pulling mainline patches from nuttx.
But libxx__gnu_unwind_find_exidx.cxx support  is added specifically support uclibc++ exception originally in nuttx

We may not need the "libxx__gnu_unwind_find_exidx.cxx " implementation itself. Let me check if i can remove both

manoharahk commented on apps/platform/gnu/Kconfig:
```diff
@@ -0,0 +1,14 @@
+# see kconfig-language at https://www.kernel.org/doc/Documentation/kbuild/kconfig-language.txt
+#
+
+config HAVE_CXXINITIALIZE
```
Ok

manoharahk commented:
Change Log:
1. Removed elf32.h and libxx__gnu_unwind_find_exidx.cxx
2. Updated CONFIG_LIBCXX in libxx/Kconfig
3. Removed H AVE_CXXINITIALIZE in libxx/Kconfig

manoharahk commented:
1. Regarding uClib++ config : It is in the existing code. No dependency with  llvm libc++. If we don't  need it anymore , we can check and remove it. Shall we remove it in new PR?

2. Regarding updating ARCHCXXINCLUDES, currently it doesn't break any builds.  Shall we do it in new PR?

seinfra commented:
Code Rule Check OK.

sunghan-chang commented:
```
Regarding uClib++ config : It is in the existing code. No dependency with llvm libc++. If we don't need it anymore , we can check and remove it. Shall we remove it in new PR?
```
I saw llvm is come from Nuttx and you use same name. But it looks strange. Config name is UCLIBCXX for uclibc++, but config name is LIBCXX for llvm, not LLVMXX. Just name side, UCLIBCXX is one of LIBCXX. That's why I want to remove UCLICXX. If we will not remove it, we had better change a name of LIBCXX. Isn't it?

```
Regarding updating ARCHCXXINCLUDES, currently it doesn't break any builds. Shall we do it in new PR?
```
I think we should include updating ARCHCXXINCLUDES in commit which you removed includes/cxx folder.
